ENH: Initial implementation of `as_euler` function.

adbugger opened the pull request:
Implemented the algorithm from the paper @nmayorov linked to. There are still some bugs to be worked out as `as_euler` is not returning the same result as input of `from_euler`. I think it has to do with the adaptations of the conventions. I'm still rechecking everything.

Meanwhile,  Nikolay if you can spot the error that would be a great help.

adbugger committed:
ENH: Initial implementation of `as_euler` function. Still buggy.

adbugger committed:
MAINT: Fixed bugs with `as_euler`. Initial implementation now works.

Had missed and arccos function previously. Understood adaptations to
algorithm. Instead of reversing the sign of every angle, we need to
reverse the axis sequence and angle sequence for extrinsic rotations.

adbugger committed:
TST: Tests for `as_euler` function

adbugger committed:
DOC: Added references

The reference links are behind a paywall. I couldn't find a link to the
full texts.

adbugger commented:
Figured out the bugs:
 - Had missed an `np.arccos` function previously
 - Had already taken a transpose of our dcm representation. But we also need to reverse angle sequence and axis sequence for extrinsic rotations.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
 AXIS_TO_IND = {'x': 0, 'y': 1, 'z': 2}
 
 
+def _elementary_basis_vector(char):
```
char -> axis

Actually I meant not using _ everywhere, but not sure.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+    # 1. Transposing our dcm representation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+
+    seq = seq[::-1] if extrinsic else seq
```
Minor, but perhaps:
```
if extrinsic:
  seq = seq[::-1]
```

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+
+def _make_euler_from_dcm(dcm, seq, extrinsic=False):
+    # The algorithm assumes intrinsic frame transformations. Their
+    # ELEMENTARY dcms are transpose of what we return.
```
"Their elementary dcms are transpose of what we return" --- this is unclear.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+    return b
+
+
+def _make_euler_from_dcm(dcm, seq, extrinsic=False):
```
Perhaps make -> compute.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+    n3 = _elementary_basis_vector(seq[2])
+
+    # Step 2
+    lamb = np.arctan2(np.dot(np.cross(n1, n2), n3), np.dot(n1, n3))
```
Let's compute `cl` and `sl` here and then call arctan2.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+    rtc = rt.dot(c)
+    ct = c.T
+    o = np.empty_like(dcm)
+    for ind in range(num_rotations):
```
Definitely we can avoid the loop here. Using `einsum` or something else.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+    angle3[adjust_mask] = (angle3[adjust_mask] - np.pi) % (2 * np.pi)
+
+    # Step 8
+    # TODO: if any observability flags are poor, possibly raise a UserWarning?
```
I believe it is a good idea. Here we have this notorious "gimbal lock", although I don't like this wording. We should say that we can't determine the angles uniquely and the third angle was set to 0 for some rotations. 

But we need to be careful, it might be not just 0 because of the adjustments coded. Needs some thinking.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,6 +8,94 @@
+
+    # Step 0
+    # Algorithms dcms are transpose of ours
+    dcm = np.transpose(dcm, (0, 2, 1))
```
As the next step I suggest to remove this transposition here. I think we can be smarter than that and show that we can adjust the algorithm accordingly without this simplest "fall back".

My idea is that if `R.T = "decomposition from the paper"`, then `R = transpose of this decomposition` and then we do small adjustments to the current code. I think with available working implementation it should be easy enough. But not until better test coverage.

adbugger replied:
I agree. I will start changing the algorithm once we're satisfied with the test coverage.

nmayorov replied:
I think we are approaching this point now.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -419,3 +507,55 @@ def from_euler(cls, seq, angles, degrees=False):
+            Adjacent axes cannot be the same.
+            Extrinsic and intrinsic rotations cannot be mixed in one function
+            call.
+
```
Without blank.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -419,3 +507,55 @@ def from_euler(cls, seq, angles, degrees=False):
+        .. [2] Malcolm D. Shuster, F. Landis Markley
+                `General Formula for Euler Angles
+                <https://arc.aiaa.org/doi/abs/10.2514/1.16622>`_
+        .. [3] Malcolm D. Shuster `A Survey of Attitude Representations
```
To be honest I'm not sure what reference [3] has to do with the conventions we use. In fact he uses the opposite convention (which I really don't like by the way).

adbugger replied:
So I was unclear here. I meant to say that the conventions used here _are_ in fact the opposite of what we use. I included it only if someone was going though reference 2 and wanted to know about the implementation.

Should I make this difference in conventions clearer or remove reference 3 completely?

nmayorov replied:
I think reference [3] is not useful for this algorithm, however it is a good reference in general, we may include it later. For now I suggest to remove it.

For the conventions used, I think our aim is too clearly explain what convention *we* use. Someone who wants to check the paper is likely to be prepared to figure out the difference between them. I think we don't need to be very verbose about that. You can mention that we use the different convention.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +455,25 @@ def test_from_euler_extrinsic_rotation_313():
+def test_as_euler_extrinsic_rotation():
+    np.random.seed(0)
+    angles = np.random.uniform(low=-np.pi, high=np.pi, size=(6, 3))
+    rot_obj = Rotation.from_euler('zxy', angles)
```
Perhaps just `rotation` or `rot`?

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +455,25 @@ def test_from_euler_extrinsic_rotation_313():
+    rot_obj = Rotation.from_euler('zxy', angles)
+
+    estimates = rot_obj.as_euler('zxy')
+    est_obj = Rotation.from_euler('zxy', estimates)
```
OK, it makes sense to me, but isn't it better to compare angles directly? Preferably we want to see, that the ranges our algorithm selects are easy to understand and intuitive. I.e if used `roll=10, pitch=20, yaw=150` we want to see the same angles back, not some bizarre transformation of them. You see my idea.

adbugger replied:
You're right. I will include some extra tests for that as well.
Since I was initializing random angles here, the returned angles may not be the same due to adjustments. Hence the dcm comparisons.

nmayorov reviewed (commented):
This is a great start and I think it's 70-80 percents work's done.

I would like to see better testing. Definitely we want to test all possible variants of axes sequences.

adbugger committed:
STY: Style changes to code

Removed leading underscores from global function names. Renamed
argument. Calculated sin(lambda) and cos(lambda) explicitly. Removed
whitespace from docstring. Modified local variable in test.

larsoner commented:
@adbugger let me know when @nmayorov is happy and I can take a look

adbugger committed:
STY: Explanation in comment is clearer

adbugger committed:
TST: Test cases for `as_euler` with extrinsic rotations

These test cases compare the actual angles returned by the algorithm. It appears that we cannot always guarantee that these will be th same as the input angles, only that the final rotation generated by the angles will be the same.

For intrinsic rotations, it appears that the algorithm never gives the same angles.

adbugger committed:
MAINT: Used np.einsum instead of loop

adbugger committed:
TST: Changed variable names in test

adbugger commented:
So I added the tests comparing angles directly. It appears that for extrinsic rotations, we cannot always guarantee that the angles returned will be the same, only that the rotation represented will be identical. However, I've added a few test cases for extrinsic rotations for which this case seems to hold.

In the intrinsic case, however, it looks like the algorithm never guarantees strict equality of angles. However, once again, the angle sequence returned represents the same rotation.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,102 @@
+    sl = np.dot(np.cross(n1, n2), n3)
+    cl = np.dot(n1, n3)
+
+    norm = sl**2 + cl**2
```
From the looks of it the norm will be one always. Check it too.

adbugger replied:
You're right. I'm not sure about the general case presented in the paper, but since we're only dealing with axes parallel to the coordinate axes, norm will always be one. Removing it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,102 @@
+    ])
+    rtc = rt.dot(c)
+    ct = c.T
+    o = np.empty_like(dcm)
```
Am I not mistaken that this line is unnecessary?

adbugger replied:
Right. Now that I've used einsum, this is not necessary.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +512,56 @@ def from_euler(cls, seq, angles, degrees=False):
+        intrinsic = (re.compile('^[XYZ]{1,3}$').match(seq) is not None)
+        extrinsic = (re.compile('^[xyz]{1,3}$').match(seq) is not None)
+        if not (intrinsic or extrinsic):
+            raise ValueError("Expected axes from `seq` to be from ['x', 'y', "
```
Slightly awkward to read, I suggest to break the line before the first list.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,102 @@
+    safe2 = (np.abs(angle2 - lamb - np.pi) >= eps)
+
+    angle1 = np.empty(num_rotations)
+    angle3 = np.empty(num_rotations)
```
Considering we use `angle2 - lamb` below, I suggest to introduce an intermediate angle as `np.arccos(o[:, 2, 2])`.

adbugger committed:
MAINT/DOC: Corrected einsum broadcasting. Added function to docs.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,102 @@
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    adjust_mask = np.logical_or(angle2 < 0, angle2 > np.pi)
```
OK, I experimented with this adjustment a little bit and I propose the following convention:

1. Second angle is from -90 to 90 degrees.
2. First and third angles are from -180 to 180 degrees. Could always achieve that, my point is not to use [0, 360], you need to modify your code a little bit.

I think this is a decent convention for common 3-1-3 and 3-2-1 (yaw, pitch, roll) sequence. At least for me it seems more natural than restricting the second angle from 0 to 180.

nmayorov replied:
Think it through or experiment --- is it possible that the second angle won't be in [-90, 90] when we apply the correction.

adbugger replied:
I was just experimenting with this and I have a question. Is is possible to represent a single rotation about, say y-axis, with angle 135 degrees, as a rotation about different axes?
If not, then `Rotation.as_euler('zyx')` cannot have angle2 in range `[-90, 90]`.

Further, for axes sequence `zxz`, parameter `lambda` is `0`. This means that adjusting `angle2` to be within range `[-90, 90]` is not possible. For example, I'm trying to do `Rotation.from_euler('zxz', [25, 135, 35], degrees=True).as_euler('zxz', degrees=True)`. Since `lambda = 0` for this, `angle2` can only be `135` or `-135` degrees. I think this is why the paper chose the range of restriction to be `[0, 180]`. For now I'm restricting range of `angle2` to be `[0, 180]` as per the paper.

adbugger replied:
Another update. So I tried this test case: `R.from_euler('zxy', [30, 100, 60], degrees=True).as_euler('zxy', degrees=True)`. The angles returned by the algorithm are `[-150.,   80., -120.]`, without any adjustments. I think we need to rethink our approach of checking for equality of angles. @nmayorov what do you suggest?

nmayorov replied:
Good investigation. I can't dig into it at the moment (will do tonight), I suggest you to consider which `phi` angles we get for different axis sequences. I will make a guess/suggestion that for `zyx` (and alike) axes we should constraint the second angle to [-90, 90] so it will behave conventionally for roll, pitch and yaw. And for 'zxz' we should constraint it into [0, 180] --- because we can't achieve [-90, 90] and it should also make sense. Please investigate it farther.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,102 @@
+    sl /= norm
+    cl /= norm
+
+    lamb = np.arctan2(sl, cl)
```
Let's use a different name, `lamb` just sounds bad. Pick another Greek letter or some informative name. You can mention that this is lambda from the paper in the comment.

nmayorov reviewed (commented):
I suggested angles convention which seems preferable to me. You are right that we can't return the input angles always, we need to establish some convention and explain it in the docstring.

nmayorov commented:
For testing:

1. Pick and implement the convention.
2. Generate many random Euler sequence which satisfy this convention. Check that the return angles match.
3. Do 2 for all possible sequences.

adbugger committed:
DOC: Removed redundant reference. Cleared up convention difference.

adbugger committed:
MAINT: Enforced angle restrictions

adbugger commented:
I have pointed out a few concerns with the angle range conventions and the checking for angle equality approach to testing.

adbugger committed:
MAINT: Updated range restrictions for second angle

adbugger committed:
TST: Use randomly generated angles for testing.

Now that the ranges for angle2 have been properly fixed, we can test for
angle equality with randomly generated angles within those ranges.

adbugger commented:
So I have enforced the angle restrictions now and changed the test cases. Now we are randomly generating angles and testing for angle equality for all allowed combinations of axes.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +456,63 @@ def test_from_euler_extrinsic_rotation_313():
     ]))
+
+
+def test_as_euler_extrinsic_rotation():
```
I would probably split tests for symmetric and asymmetric axis sequence (i.e. zxz and zyx types). And test for intrinsic and extrinsic rotations at the same place.

For zyx type permutations are correct approach, but for zxz perhaps just use two nested loops.

My point is that the code could be a little bit more readable.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +518,58 @@ def from_euler(cls, seq, angles, degrees=False):
+        extrinsic = (re.compile('^[xyz]{1,3}$').match(seq) is not None)
+        if not (intrinsic or extrinsic):
+            raise ValueError("Expected axes from `seq` to be from "
+                             "['x', 'y', 'z'] or "
```
Can we put ['X', 'Y', 'Z'] on this line as well?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,107 @@
+
+    # angle phi is lambda from the paper referenced in [2] from docstring of
+    # `as_euler` function
+    phi = np.arctan2(sl, cl)
```
I believe `offset` might be just the right name for it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,107 @@
+    o = np.einsum('...ij,...jk->...ik', res, ct)
+
+    # Step 4
+    offset = np.arccos(o[:, 2, 2])
```
Really preferential, but maybe call it `angle2` and then update it as `angle2 += offset` (assuming `phi` will be named as `offset`).

adbugger replied:
We need the intermediate quantity `np.arccos(o[:, 2, 2])` (which was `offset` before, now renamed to `phi`) in the condition checks below.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,107 @@
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angle2 < 0, angle2 > np.pi)
+    else:
+        adjust_mask = np.logical_or(angle2 < -np.pi / 2, angle2 > np.pi / 2)
```
Fo this case, what are possible values for `phi`? +pi/2 or -pi/2?

adbugger replied:
Yes, I'll put that as a comment.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,107 @@
+    angle2[adjust_mask] = 2 * phi - angle2[adjust_mask]
+    angle3[adjust_mask] = angle3[adjust_mask] - np.pi
+
+    angles = np.column_stack((angle1, angle2, angle3))
```
Just an idea: perhaps introduce `angles` array from the beginning and work this it instead of `angle1`, `angle2`, `angle3`? Not sure.

adbugger replied:
I guess this makes sense now that we're making adjustments as a whole.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +518,58 @@ def from_euler(cls, seq, angles, degrees=False):
+        opposed to rotation matrices. Thus, the matrix representation used in
+        the paper is a transpose of the direction cosine matrix representation
+        returned by the `as_dcm` function.
+
```
Definitely should write about ranges of the returned angles (we have two cases now).

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,107 @@
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    # TODO: if any observability flags are poor, possibly raise a UserWarning?
```
As I see it the degenerate angle might be only zero, because the troublesome value of the second angle will be right on the boundary of the valid range, so it won't be corrected. 

I see two approaches:

1. Issue a warning and explain that such case is possible in the docstring.
2. Not issue a warning but explain that such case is possible in the docstring.

I lean towards the second option.

adbugger replied:
[Commenting here because the other one is being collapsed]
So I tested dcm equality for the degenerate case. The dcms are not equal upto 6 decimal places. Since it appears that the algorithm is not numerically stable for this case, will it be better to raise a Warning / Error? Or do we look for a better algorithm?

nmayorov replied:
I don't think there is a better algorithm. Now as you said it, a warning looks appropriate. I will investigate this point myself tonight.

adbugger replied:
One particularly tricky case of gimbal lock I found was `Rotation.from_euler('yzx', [45, 90, 35], degrees=True)`. For this, `lambda = -90.0 degrees`, and all observability flags are `True`. Also note that the third angle returned was not `0`. The current implementation of `as_euler('yzx', degrees=True)` returns `[36.86989765, 89.99999879, 30.25643716]`.

It appears that we cannot detect all possible cases of gimbal lock by relying on the algorithm in the paper. I think that the paper's observability flags are not directly related to gimbal lock at all. This needs some discussion.

nmayorov replied:
I checked this case and I think I figured the problem. 

We have `dcm_transformed[:, 2, 2]` very close to -1 (should be -1 exactly) and then we have `np.arccos(dcm_transformed[:, 2, 2])` different from `pi` by something about `1e-8`. Speaking in exact numbers we definitely have the gimbal lock here. The problem occurs because you chose epsilon too small. By the way, why you chose it that way? Because of that we use "normal" formulas and get quite weird result.

I suggest to tune this eps such as to avoid such strange results for all sequences and when we set the second angle to create the gimbal lock situation (using degrees).

adbugger replied:
I thought choosing `eps` to be at machine precision would be fine. I'll tune it.

nmayorov replied:
It is quite tricky and I'm not sure exactly how to analyze it. But your example showed that if we don't consider this case at all or choose eps too small then we will have bad results. Here we have `36.86989765 + 30.25643716 = 67.12633481` and the correct angles should add up to 80. In other words we got quite inaccurate dcm.

nmayorov replied:
Also I notices that for extrinsic angles we will have the first angle zero, not the third. Should we change it?

nmayorov replied:
Another thing I noticed: we probably should not apply `adjust_mask` for "gimbal lock cases". Try `Rotation.from_euler('YZX', [45, 90, 35], degrees=True)` to see what happens. The problem we have is that the second angle might be slightly off the desired boundary (less than epsilon), basically it means that this angle is very close to the critical value and we declare the "gimbal lock case" in our numerical algorithm. In other words we assume that this angle is equal to the gimbal lock angle exactly, it seems that we should not adjust it.

adbugger replied:
So I put `eps = 1e-7`, since that was the smallest value that passes the test cases. I think it is fine since `assert_allclose` also has an absolute tolerance of 6 decimal places, but I want to know your opinion on this.

adbugger replied:
> Also I notices that for extrinsic angles we will have the first angle zero, not the third. Should we change it?

I'm sorry change what exactly? Right now I'm detecting gimbal lock cases by the 'observability flag' mentioned in the paper.

adbugger replied:
> Another thing I noticed: we probably should not apply adjust_mask for "gimbal lock cases". Try Rotation.from_euler('YZX', [45, 90, 35], degrees=True) to see what happens. The problem we have is that the second angle might be slightly off the desired boundary (less than epsilon), basically it means that this angle is very close to the critical value and we declare the "gimbal lock case" in our numerical algorithm. In other words we assume that this angle is equal to the gimbal lock angle exactly, it seems that we should not adjust it.

I see your point. For your example we get `[-100.,  89.99999879, -180.]` after adjustment. I'll remove the application of `adjust mask` for these cases. Our range restrictions may not apply for these cases but I'm not too worried since even in these cases our ranges will be off by at most `eps`.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +456,63 @@ def test_from_euler_extrinsic_rotation_313():
+        assert_allclose(
+            angles_same_axes,
+            Rotation.from_euler(
+                seq, angles_same_axes).as_euler(seq))
```
Besides those tests you definitely should test for degenerate case, where the first and third angles can't be determined uniquely.

adbugger replied:
So I tested dcm equality for the degenerate case. The dcms are not equal upto 6 decimal places. Since it appears that the algorithm is not numerically stable for this case, will it be better to raise a `Warning` / `Error`? Or do we look for a better algorithm?

nmayorov reviewed (commented):
I think we are close, need to clean up some things and add some tests.

adbugger committed:
TST: Split tests according to axes type, symmetric and asymmetric

adbugger committed:
MAINT: Renamed variables appropriately

adbugger committed:
DOC: Clarified angle ranges in the docstring

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,106 @@
+    o = np.einsum('...ij,...jk->...ik', res, ct)
+
+    # Step 4
+    phi = np.arccos(o[:, 2, 2])
```
I suggested to assign angles column right here, without introducing phi. Then update it as `angles[:, 1] += offset`, when we can. I think now that it is probably worse, decide yourself.

adbugger replied:
Oh, I understand now. Making the change.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,106 @@
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angles[safe_mask, 0] = np.arctan2(o[safe_mask, 2, 0], -o[safe_mask, 2, 1])
```
I think `o` is a bit inexpensive name for a variable, think about something else. Like dcm_transformed perhaps.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,106 @@
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    ct = c.T
```
Could live without introducing ct.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +517,67 @@ def from_euler(cls, seq, angles, degrees=False):
+            - Third angle belongs to [-180, 180] degrees (both inclusive)
+            - Second angle belongs to:
+
+                - [0, 180] degrees if all axes are unique (like xyz)
```
I think it is vice-versa.

nmayorov reviewed (commented):
Few notes. Was writing from the phone, will get back to it tonight.

adbugger committed:
STY/DOC: Style changes to code, fixed error in docstring

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +524,67 @@ def from_euler(cls, seq, angles, degrees=False):
+            - Second angle belongs to:
+
+                - [-90, 90] degrees if all axes are unique (like xyz)
+                - [0, 180] degrees if all axes are not unique (like zxz)
```
I would write "if the thirst and the third axes are the same". And in the first case "all axes are different. And probably add single quotes ('xyz').

By the way, have you checked that this is rendered nicely?

adbugger replied:
![image](https://user-images.githubusercontent.com/15844796/40881904-4d9ef2da-66f0-11e8-876c-d44f6774612b.png)

Yup, it renders fine on my setup.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -8,7 +8,113 @@
+    # Step 0
+    # Algorithms dcms are transpose of ours
+    dcm = np.transpose(dcm, (0, 2, 1))
+    angles = np.empty((num_rotations, 3))
```
Create `angles` right before the first assignment for it.

adbugger committed:
TST: Add tests to ensure proper handling of gimbal lock

adbugger committed:
MAINT: Fixed detection of gimbal lock

The problem was that the value of eps had been taken to be too small. It
is now the smallest value that passes the tests. (~1e-7). Also make some
minor style and doc changes.

adbugger committed:
DOC: Mentioned behaviour for cases of gimbal lock in docstring

adbugger committed:
TST: Tests for gimbal lock cases

adbugger committed:
MAINT: Don't apply angle adjustment for gimbal locked cases

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,127 @@
+    # angle offset is lambda from the paper referenced in [2] from docstring of
+    # `as_euler` function
+    offset = np.arctan2(sl, cl)
+    c = np.empty((3, 3))
```
Up to you, but in this case it looks OK to use `np.vstack`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,127 @@
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
```
I think it's OK to move the computation to the next line and don't introduce a new variable.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,127 @@
+    )
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
```
The comment is obsolete.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,127 @@
+                                    angles[:, 1] > np.pi / 2)
+
+    # Dont adjust gimbal locked angle sequences
+    adjust_mask = np.logical_and(adjust_mask, safe_mask)
```
Do you prefer using this instead of `&`? I'm totally fine with it, just asking.

adbugger replied:
I honestly wasn't aware of this symbol for element-wise `and`. The first Google search result for 'python element wise and' led me to `numpy.logical_and`.
Do you want me to change it?

nmayorov replied:
As you prefer, I'm fine with the current code.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,127 @@
+    if not np.all(safe_mask):
+        warnings.warn("Some angle sequences suffer from gimbal lock. In those "
+                      "cases, it is not possible to determine the angles "
+                      "uniquely.", UserWarning, stacklevel=1)
```
Perhaps add "the third angle was set to 0". Also what about those setting `UserWarning, stackleve=1`? Why you decided to use these?

If possible, could you try to shorten the warning message? I think shorter the message --- easier for a user to quickly read it and understand.

adbugger replied:
`UserWarning` is the default warning class. I decided to state that because I'm checking for that particular class in the tests.
Regarding the `stacklevel=1`, I thought it would be appropriate if the warning was raised at the `as_euler` function. Maybe I'm wrong here.

I'll remove the stack level and add the message about the third angle.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +533,74 @@ def from_euler(cls, seq, angles, degrees=False):
+            - Third angle belongs to [-180, 180] degrees (both inclusive)
+            - Second angle belongs to:
+
+                - [-90, 90] degrees if all axes are unique (like xyz)
```
I suggest unique -> different, xyz->'xyz' and same -> the same.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +533,74 @@ def from_euler(cls, seq, angles, degrees=False):
+                - [-90, 90] degrees if all axes are unique (like xyz)
+                - [0, 180] degrees if first and third axes are same (like zxz)
+
+        Euler angles suffer from the problem of gimbal lock. It occurs when,
```
Add a reference to here https://en.wikipedia.org/wiki/Gimbal_lock#In_applied_mathematics

I believe we can omit the explanation and rely on the reference. However, I suggest to write ("In this case the third angle is set to zero. Note however that the returned angles still represent the correct rotation".

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,127 @@
+                      "uniquely.", UserWarning, stacklevel=1)
+
+    # Reverse role of extrinsic and intrinsic rotations
+    return angles[:, ::-1] if extrinsic else angles
```
In case of gimbal lock we will have for example [10, 90, 0] and then after reversion [0, 90, 10]. My idea/suggestion is to perhaps always return the third angle as zero for intrinsic and extrinsic sequences.

I think you can simply swap angles 1 and 3 for gimbal lock cases in extrinsic sequences. What do you think?

nmayorov replied:
Looks like you can write 
```
if extrinsic:
   angles[safe_mask] = angles[safe_mask, ::-1]
```

adbugger replied:
So doing this causes the tests to fail. I inspected the dcm of the returned and it looks like not reversing the angle sequences leads to sign reversals of `dcm[:, 0, 1]` and `dcm[:, 1, 2]`. Both of these quantities are small and close to zero, but the change in sign leads to assertion failure.

What do you suggest? Should we say that the first angle is zero for extrinsic rotations?

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +456,122 @@ def test_from_euler_extrinsic_rotation_313():
+    # Since we cannot check for angle equality, we check for dcm equality
+    angles = np.array([
+        [45, 90, 35],
+        [90, 35, 20],
```
A bit confused, why do you believe [90, 35, 20] creates gimbal lock situation?

I thought it occurs only when the second angle is 90 or -90.

adbugger replied:
Ah, you're absolutely right. My mistake.

nmayorov reviewed (commented):
A bit confused about new tests.

adbugger committed:
TST: Gimbal lock can only be caused by second angle being close to 90

adbugger committed:
DOC: Removed explanation, added reference for gimbal lock

adbugger committed:
STY: Minor changes to code style

adbugger committed:
MAINT: Handling Gimbal lock.

Clarified warning for gimbal lock. Also tried not to reverse angle
sequences for gimbal locked cases but that leads to test failures.

adbugger commented:
[commenting here because the other thread is collapsed]

So not reversing angle sequence for gimbal locked cases causes the tests to fail. I inspected the dcm of the returned and it looks like not reversing the angle sequences leads to sign reversals of dcm[:, 0, 1] and dcm[:, 1, 2]. Both of these quantities are small and close to zero, but the change in sign leads to assertion failure.

What do you suggest? Should we say that the first angle is zero for extrinsic rotations?

nmayorov commented:
About gimbal lock and extrinsic rotations. You are right that we can not just reassign the first angle to third, however there is exists a proper solution. Check formulas 11a and 11b, you can set phi to zero from them and compute psi. I think it's better to introduce two if branches for 6a-6c steps ins your code: one that is already coded and another for extrinsic rotations.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +456,122 @@ def test_from_euler_extrinsic_rotation_313():
+        [45, 90, 35],
+        [35, -90, 20],
+        [35, 90, 25],
+        [15, 0, 60],
```
I believe when the second angle is zero we don't have the gimbal lock (for xyz sequences).

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -455,3 +456,122 @@ def test_from_euler_extrinsic_rotation_313():
+def test_as_euler_degenerate_symmetric_axes():
+    # Since we cannot check for angle equality, we check for dcm equality
+    angles = np.array([
+        [45, 90, 35],
```
Here 90 degrees does not create the gimbal lock. However 180 degrees will --- add some of those.

adbugger committed:
MAINT: Handle extrinsic rotations in a separate branch

This is to ensure that after reversing angle sequence we still get third
angle as zero

adbugger committed:
TST: Angles specific to gimbal lock cases

adbugger committed:
MAINT: Remove dcm transposition and adapt rest of algorithm

adbugger commented:
So I handled extrinsic rotations in another branch. After some experimentation, I realized I had to interchange the role of equations `11a` and `11b` as well. 

I also removed the dcm transposition in the beginning. Now, `dcm_transformed` is the transpose of what it previously was and we're swapping the second and third indices wherever applicable. 

I also updated the test cases according to your suggestions and they pass.

adbugger committed:
DOC: Fixed docstring

nmayorov commented:
Looks excellent to me!

@larsoner take a look as well.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,139 @@
 
 
-def _make_elementary_quat(axis, angles):
+def elementary_basis_vector(axis):
```
Anything that we do not to support the user using (e.g., no docstring, not in docs, etc.) should be preceded by a leading `_` to make it private rather than public -- that is a signal to users they should not expect API stability if they go browsing the `scipy.spatial.transform.rotation` namespace. In general you should restrict the set of public functions as much as possible to reduce maintenance burden and keep the code as flexible as possible for future updates.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,139 @@
+    return b
+
+
+def compute_euler_from_dcm(dcm, seq, extrinsic=False):
```
Make private

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -3,12 +3,139 @@
+    return angles
+
+
+def make_elementary_quat(axis, angles):
```
Make private

I'll stop commenting on each of these, but I assume you get the idea.

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -28,18 +155,18 @@ def _compose_quat(p, q):
 
 
-def _elementary_quat_compose(seq, angles, intrinsic=False):
+def elementary_quat_compose(seq, angles, intrinsic=False):
```
leave as private

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -417,5 +545,74 @@ def from_euler(cls, seq, angles, degrees=False):
+        if len(seq) != 3:
+            raise ValueError("Expected 3 axes, got {}.".format(seq))
+
+        intrinsic = (re.compile('^[XYZ]{1,3}$').match(seq) is not None)
```
Any reason not to use `re.match` directly?

larsoner reviewed (commented):
Otherwise LGTM

nmayorov commented:
Hey Eric, it was my suggestion (sort of) not to use underscores. My motivation was that rotation.py is not a public module and we can control what to import from it to transform in __init__.

Should we write _rotation.py as well? I'm a bit confused with the levels of private modifiers.

larsoner commented:
> Should we write _rotation.py as well? I'm a bit confused with the levels of private modifiers.

You could, sure. Leaving it as `rotation.py` means someone could do `from scipy.spatial.transform import rotation` without having to pass any `_`, which makes it seem public (even if it's not in the docs). But then again if all of the functions in there have a preceding `_` (or are public elsewhere) then it doesn't matter much either way.

There is an additional maintainability benefit to having leading `_` on functions and classes, namely that the next contributor immediately knows it is not public. Even if you name the file `_rotation.py`, some `my_func` inside it could be imported in `scipy/spatial/transforms/__init__.py` as `from ._rotation import my_func`, so seeing `_rotation.py` is no guarantee that everything inside it is private.

nmayorov commented:
So your overall suggestion is to do both: use `_rotation.py` and mark everything non-public inside with `_` as well? Did I understand correctly?

adbugger committed:
MAINT: Add leading '_' to private functions

adbugger commented:
Added leading `_` to make functions private. Will wait for confirmation before renaming file.

larsoner commented:
I'm happy either way on the filenames
